### PR TITLE
function isValidReturnType added

### PR DESCRIPTION
**Pull request by Hatsunespica**
Check if the given OperationOp can use specified type as a valid return type.
`bool isValidReturnType(GeneratorInfo &info, irdl::OperationOp op, mlir::Type type)`

For each constraint variable, we try to assign the constraint with the provided type. Otherwise use random one from all satisfied types.

### CODE REVIEW

File: mlir-enumerate/mlir-enumerate.cpp [hunk 1]
```diff
@@ -247,6 +247,73 @@ LogicalResult addOperation(GeneratorInfo &info) {
...
+              .cast<TypeConstraintAttrInterface>()
+              .getTypeConstraint(info.irdlContext, namedConstraintVars);
+      // TODO(fehr) Currently a hack, will be fixed later once I update
+      // the IRDL API.
+      auto constraint2 =
+          namedConstraint.getConstraint()
+              .cast<TypeConstraintAttrInterface>()
+              .getTypeConstraint(info.irdlContext, namedConstraintVars);
+
+      if (constraint.get()
```
Comment (math-fehr) [inline mlir-enumerate/mlir-enumerate.cpp]: You do not actually need to assign any type on the constraints here.
This should be done automatically when you check that the result types are satisfying the constraints.
So when you check something like `Vector<T>`  with the type `Vector<i32>` for instance, `T` will be set to `i32` in the `verifyType` function.

So for here, you can actually just check the result constraints, without dealing manually with the constraint variables.
You still though have to fill `namedConstraintVars`, `varConstraints`, and `vars`. But you should just set each `Type` to `Type()` in `vars` if I recall correctly.

**Comment by Hatsunespica** [comment]
Hi Mathieu,

Thanks for your reviewing!
I’ll update my code like you suggested 0w0.

All the best,
Yuyou

**Comment by Hatsunespica** [comment]
Hi Mathieu,

I updated the branch with what you said!
Does it look good?
0w0

**Comment by math-fehr** [comment]
Perfect, thanks a lot!

**Pull request merged**